File: android/app/src/main/java/dev/idank/whisper/services/WebsocketService.java
package dev.idank.whisper.services;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import dev.idank.whisper.R;
import dev.idank.whisper.clients.WebsocketClient;

public class WebsocketService extends Service {

    private final ExecutorService executorService = Executors.newSingleThreadExecutor();
    private WebsocketClient wsClient;

    @Override
    public void onCreate() {
        super.onCreate();
        startForeground(1, createNotification());
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        if (intent != null && wsClient == null) {
            String apiToken = intent.getStringExtra("apiToken");
            String websocketURL = intent.getStringExtra("websocketURL");
            String deviceID = intent.getStringExtra("deviceID");

            WebsocketClient.setInstance(this, apiToken, websocketURL, deviceID);
            wsClient = WebsocketClient.getInstance();

            executorService.submit(wsClient::connect);
        }

        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (wsClient != null) {
            executorService.submit(wsClient::shutdown);
            executorService.shutdown();
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private Notification createNotification() {
        String channelId = "whisper_service_channel";

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    channelId,
                    "Service Channel",
                    NotificationManager.IMPORTANCE_LOW
            );
            getSystemService(NotificationManager.class).createNotificationChannel(channel);
        }

        return new NotificationCompat.Builder(this, channelId)
                .setContentTitle("Service Running")
                .setContentText("Background task active")
                .setSmallIcon(R.drawable.ic_launcher_foreground)
                .build();
    }
}

File: android/app/src/main/java/me/idank/whisper/services/WebSocketService.kt
package me.idank.whisper.services

import android.R
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.os.IBinder
import androidx.core.app.NotificationCompat
import me.idank.whisper.managers.WebSocketManager

class WebSocketService : Service() {

    private lateinit var manager: WebSocketManager

    override fun onCreate() {
        super.onCreate()
        startForegroundServiceNotification()

        manager = WebSocketManager.getInstance()
        manager.connectWithContext(this)
    }

    private fun startForegroundServiceNotification() {
        val channelId = "websocket_channel"
        val channel = NotificationChannel(
            channelId,
            "WebSocket Service",
            NotificationManager.IMPORTANCE_LOW
        )
        (getSystemService(NOTIFICATION_SERVICE) as NotificationManager)
            .createNotificationChannel(channel)

        val notification = NotificationCompat.Builder(this, channelId)
            .setContentTitle("Whisper WebSocket Connected")
            .setSmallIcon(R.drawable.stat_sys_download_done)
            .build()

        startForeground(1, notification)
    }

    override fun onDestroy() {
        super.onDestroy()
        manager.disconnect()
    }

    override fun onBind(intent: Intent?): IBinder? = null
}
